File: server/src/main/java/rikka/shizuku/server/PermissionObserver.java
package rikka.shizuku.server;

import android.content.Context;
import android.content.pm.PackageManager;
import java.lang.reflect.Proxy;
import java.util.function.IntConsumer;

final class PermissionObserver {
    static void register(IntConsumer changed) throws ReflectiveOperationException {
        Class<?> activityThreadClass = Class.forName("android.app.ActivityThread");
        Object thread = activityThreadClass.getMethod("currentActivityThread").invoke(null);
        if (thread == null) thread = activityThreadClass.getMethod("systemMain").invoke(null);
        Context context = (Context) activityThreadClass.getMethod("getSystemContext").invoke(thread);
        Class<?> listenerClass = Class.forName("android.content.pm.PackageManager$OnPermissionsChangedListener");
        Object listener = Proxy.newProxyInstance(listenerClass.getClassLoader(), new Class<?>[]{listenerClass},
                (proxy, method, args) -> {
                    switch (method.getName()) {
                        case "onPermissionsChanged": changed.accept((Integer) args[0]); return null;
                        case "hashCode": return System.identityHashCode(proxy);
                        case "equals": return proxy == args[0];
                        case "toString": return "Porter permission observer";
                        default: return null;
                    }
                });
        PackageManager.class.getMethod("addOnPermissionsChangeListener", listenerClass)
                .invoke(context.getPackageManager(), listener);
    }

    private PermissionObserver() {}
}

File: server/src/main/java/eu/darken/porter/privileged/PermissionObserver.kt
package eu.darken.porter.privileged

import android.content.Context
import android.content.pm.PackageManager
import java.lang.reflect.Proxy

internal object PermissionObserver {

    @Throws(ReflectiveOperationException::class)
    fun register(changed: (Int) -> Unit) {
        val activityThreadClass = Class.forName("android.app.ActivityThread")
        var thread = activityThreadClass.getMethod("currentActivityThread").invoke(null)
        if (thread == null) thread = activityThreadClass.getMethod("systemMain").invoke(null)
        val context = activityThreadClass.getMethod("getSystemContext").invoke(thread) as Context
        val listenerClass = Class.forName("android.content.pm.PackageManager\$OnPermissionsChangedListener")
        val listener = Proxy.newProxyInstance(listenerClass.classLoader, arrayOf(listenerClass)) { proxy, method, args ->
            when (method.name) {
                "onPermissionsChanged" -> {
                    changed(args[0] as Int)
                    null
                }
                "hashCode" -> System.identityHashCode(proxy)
                "equals" -> proxy === args[0]
                "toString" -> "Porter permission observer"
                else -> null
            }
        }
        PackageManager::class.java.getMethod("addOnPermissionsChangeListener", listenerClass)
            .invoke(context.packageManager, listener)
    }
}
